### PR TITLE
Allow "+" character in usernames

### PR DESCRIPTION
**Pull request by pako81**
## Description

Allow `+` character when creating users.

## Related Issue

- Relates to https://github.com/owncloud/guests/pull/384
- Fixes https://github.com/owncloud/enterprise/issues/3712

## Motivation and Context

This PR allows the `+` character in the username when creating new users. This is also required when sharing with guest users having i.e. the following format `name+2@test.com`.

## How Has This Been Tested?

Manually by creating new users having the `+` sign in their usernames over the Users management page as well by using the `user:add` occ command.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised 
- [x] Changelog item

### CODE REVIEW

File: lib/private/User/Manager.php [hunk 1]
```diff
@@ -332,10 +332,10 @@ public function createUser($uid, $password) {
 			$l = \OC::$server->getL10N('lib');
 
 			// Check the name for bad characters
-			// Allowed are: "a-z", "A-Z", "0-9" and "_.@-'"
-			if (\preg_match('/[^a-zA-Z0-9 _\.@\-\']/', $uid)) {
+			// Allowed are: "a-z", "A-Z", "0-9" and "+_.@-'"
+			if (\preg_match('/[^a-zA-Z0-9 \+_\.@\-\']/', $uid)) {
```
Comment (phil-davis) [inline lib/private/User/Manager.php]: Note: the escaping of `\+` here is not strictly required (just like the escaping of `\.` is actually not strictly required) but it does not hurt. In other contexts in a regex the plus sign does need to be escaped, so escaping it here just helps people not to have to think about that.

**Comment by phil-davis** [comment]
@pako81 I adjusted acceptance tests to include some cases with a user with `+` in the username. And also added a changelog entry.

IMO  this works. @micbar this could be merged for 10.4 - is that OK?

**Comment by pako81** [comment]
thanks a lot @phil-davis :)

**Comment by pako81** [comment]
@phil-davis I was now thinking about the user provisioning APIs. In there, the `+` sign is still not correctly escaped and rather replaced with a space character.

**Comment by phil-davis** [comment]
@pako81 I added provisioning API tests. They are passing locally for me and I expect that they will pass in drone CI. 
Probably the tests are providing the `+` in the URL as ASCII 43, `%2B` (I think that is true, without actually digging deep into the test code)
And that is how any client using the Provisioning API would need to specify a username containing a `+` - and the backend code will receive the `%2B` turned back into a `+` and it "just works"

**Comment by pako81** [comment]
@phil-davis correct, I was testing the wrong way. The following seems to work correctly.

`curl -X POST http://admin:password@example.com/ocs/v1.php/cloud/users -d userid="name%2B1@test.com" -d password="test"`

**Comment by phil-davis** [comment]
There have been other major features merged just now. So I rebased this PR to make sure that the CI is all green on top of those changes now in master.

**Comment by phil-davis** [comment]
@micbar @HanaGemela I have added acceptance test coverage and it passes (adding users with `+` in ther username via the webUI, cli and provisioning API and doing stuff with them...) So this looks like it works fine.

I some developer can review the code and approve, and you think this is a change/enhancement that we should have, then feel free to merge it for 10.4

**Review by sharidas — COMMENTED** [review]
LGTM :+1:

**Review by karakayasemi — CHANGES_REQUESTED** [review]
Everything seems good. As a last, this should be adjusted also.
https://github.com/owncloud/core/blob/98922f11f87cb4818212d530a59c51d57fb4f124/core/Command/User/Add.php#L67

**Comment by pako81** [comment]
thanks @karakayasemi :) missed that

**Review by karakayasemi — APPROVED** [review]

**Pull request merged**